### PR TITLE
CTX-5758: Improved autoupdate implementation.

### PR DESCRIPTION
**Pull request by bogdant36**
CTX-5758: Changed implementation to call "node update" command with --auto flag instead of running python scripts or anything similar.

### CODE REVIEW

File: coretex/cli/commands/node.py [hunk 1]
```diff
@@ -117,8 +125,8 @@ def update() -> None:
 
     node_module.stop()
 
+    stdEcho("Updating node...")
     node_module.start(dockerImage, config)
-    docker.removeDanglingImages(node_module.getRepoFromImageUrl(dockerImage), node_module.getTagFromImageUrl(dockerImage))
```
Comment (dule1322) [inline coretex/cli/commands/node.py:121]: Why was this removed?

File: coretex/utils/docker.py [hunk 2]
```diff
@@ -3,6 +3,10 @@
 import json
 
 from .process import command, CommandException
+from ..configuration import loadConfig
+
+
+config = loadConfig()
```
Comment (dule1322) [inline coretex/utils/docker.py:9]: Is this intended?

File: coretex/cli/modules/utils.py [hunk 3]
```diff
@@ -19,12 +19,21 @@
...
 from py3nvml import py3nvml
 
 import click
 
 from ...utils.process import command
+from ...configuration import DEFAULT_VENV_PATH
+
+
+def checkEnvironment() -> None:
+    if not DEFAULT_VENV_PATH.exists():
```
Comment (dule1322) [inline coretex/cli/modules/utils.py:33]: Invert the check

File: coretex/cli/modules/utils.py [hunk 4]
```diff
@@ -19,12 +19,21 @@
...
 
 import sys
+import venv
 
 from py3nvml import py3nvml
 
 import click
 
 from ...utils.process import command
+from ...configuration import DEFAULT_VENV_PATH
```
Comment (dule1322) [inline coretex/cli/modules/utils.py:29]: Sort imports properly

File: coretex/cli/modules/utils.py [hunk 5]
```diff
@@ -19,12 +19,21 @@
...
+
+
+def checkEnvironment() -> None:
+    if not DEFAULT_VENV_PATH.exists():
+        venv.create(DEFAULT_VENV_PATH, with_pip = True)
+        venvPython = DEFAULT_VENV_PATH / "bin" / "python"
+        command([str(venvPython), "-m", "pip", "install", "/Users/bogdanbm/Documents/working/biomech/source/coretexpylib"])
```
Comment (dule1322) [inline coretex/cli/modules/utils.py:36]: Local path for coretexpylib is hardcoded. Get the installation source properly and use that here as well.

File: coretex/cli/modules/update.py [hunk 6]
```diff
@@ -41,51 +31,9 @@ class NodeStatus(IntEnum):
...
-
-
-def activateAutoUpdate(configDir: Path, config: Dict[str, Any]) -> None:
-    updateScriptPath = CONFIG_DIR / UPDATE_SCRIPT_NAME
-    dumpScript(updateScriptPath, config)
-
-    if not jobExists(UPDATE_SCRIPT_NAME):
-        scheduleJob(configDir, UPDATE_SCRIPT_NAME)
+def activateAutoUpdate() -> None:
+    if not jobExists('coretex node update --auto'):
```
Comment (dule1322) [inline coretex/cli/modules/update.py:35]: Use double quotes instead of single quotes

File: coretex/cli/modules/cron.py [hunk 7]
```diff
@@ -36,15 +36,21 @@ def jobExists(script: str) -> bool:
...
+
+    dockerPathParts = dockerPath.strip().split('/')
+    dockerExecPath = '/'.join(dockerPathParts[:-1])
+
+    gitPathParts = gitPath.strip().split('/')
+    gitExecPath = '/'.join(gitPathParts[:-1])
 
     existingLines = getExisting()
-    cronEntry = f"*/30 * * * * {configDir / script}\n"
+    cronEntry = f"PATH={gitExecPath}:{dockerExecPath}\n*/5 * * * * {DEFAULT_VENV_PATH}/bin/coretex node update --auto >> {DEFAULT_VENV_PATH.parent}/out.txt  2>&1\n"
```
Comment (dule1322) [inline coretex/cli/modules/cron.py:50]: `PATH` should not be modified as it affects all cron jobs

**Review by dule1322 — CHANGES_REQUESTED** [review]

File: coretex/cli/modules/utils.py [hunk 8]
```diff
@@ -15,18 +15,28 @@
...
 import click
 
+from ...configuration import DEFAULT_VENV_PATH
 from ...utils.process import command
 
 
+def checkEnvironment() -> None:
+    if DEFAULT_VENV_PATH.exists():
+        return
+    venv.create(DEFAULT_VENV_PATH, with_pip = True)
```
Comment (dule1322) [inline coretex/cli/modules/utils.py:35]: Add newline above 35th line

File: coretex/cli/modules/utils.py [hunk 9]
```diff
@@ -15,18 +15,28 @@
...
+    venvPython = DEFAULT_VENV_PATH / "bin" / "python"
```
Comment (dule1322) [inline coretex/cli/modules/utils.py:36]: This will not work on windows as it has different structure for venv

File: coretex/cli/resources/update_node.py [hunk 10]
```diff
@@ -0,0 +1,42 @@
...
+
+from coretex.utils import docker
+from coretex.cli.modules import node as node_module
+from coretex.cli.modules.update import getNodeStatus, NodeStatus
+from coretex.configuration import loadConfig
+
+
+autoUpdateLogger = logging.getLogger("cli-autoupdate")
+
+def updateNode() -> None:
```
Comment (dule1322) [inline coretex/cli/resources/update_node.py:11]: Add one more newline above function

File: coretex/cli/resources/update_node.py [hunk 11]
```diff
@@ -0,0 +1,42 @@
+import logging
+
+from coretex.utils import docker
+from coretex.cli.modules import node as node_module
+from coretex.cli.modules.update import getNodeStatus, NodeStatus
+from coretex.configuration import loadConfig
+
+
+autoUpdateLogger = logging.getLogger("cli-autoupdate")
```
Comment (dule1322) [inline coretex/cli/resources/update_node.py:9]: To which file does this logger output logs?

**Review by dule1322 — CHANGES_REQUESTED** [review]

File: coretex/cli/modules/utils.py [hunk 12]
```diff
@@ -15,18 +15,34 @@
...
+    if DEFAULT_VENV_PATH.exists():
+        return
+
+    venv.create(DEFAULT_VENV_PATH, with_pip = True)
+
+    if platform.system() == "Windows":
+        venvPython = DEFAULT_VENV_PATH / "Scripts" / "python.exe"
+
+    command([str(venvPython), "-m", "pip", "install", "coretex"])
```
Comment (dule1322) [inline coretex/cli/modules/utils.py:43]: This will only install production coretexpylib. It should use the same source of installation which was used for installing the CLI.

File: coretex/cli/modules/cron.py [hunk 13]
```diff
@@ -36,15 +37,12 @@ def jobExists(script: str) -> bool:
...
 
 
-def scheduleJob(configDir: Path, script: str) -> None:
-    if jobExists(script):
-        return
-
+def scheduleJob(scriptName: str) -> None:
     existingLines = getExisting()
-    cronEntry = f"*/30 * * * * {configDir / script}\n"
+    cronEntry = f"*/30 * * * * {DEFAULT_VENV_PATH}/bin/python {RESOURCES_DIR}/{scriptName} >> {DEFAULT_VENV_PATH.parent}/out.txt  2>&1\n"
```
Comment (dule1322) [inline coretex/cli/modules/cron.py:42]: Pipe the output to a proper logs directory

Comment (dule1322) [inline coretex/cli/modules/cron.py:42]: Why don't we make it so that it sources the environment and then runs "coretex node update" command directly and set a flag for it to reject all prompts like the one we have where when the command is executed if the Node is busy and it prompts user if they want to stop the Node to update it. We can add a flag which automatically anwers "No" to these questions.

File: coretex/cli/resources/update_script_template.sh [hunk 14]
```diff
@@ -1,219 +1,14 @@
...
-        echo "Node failed to update with exit code $exit_code"
-    fi
-}}
+# Set PATH to include the directory where docker is located
+export PATH=$DOCKER_PATH:$GIT_PATH
 
-# Main execution
-run_node_update
+# Execute the coretex command
+$VENV_PATH/bin/coretex node update --auto >> /Users/bogdanbm/.config/coretex/logs/ctx_autoupdate.log 2>&1
```
Comment (dule1322) [inline coretex/cli/resources/update_script_template.sh:14]: path is hardcoded

File: coretex/cli/commands/node.py [hunk 15]
```diff
@@ -73,8 +73,9 @@ def stop() -> None:
 
 
 @click.command()
+@click.option("--auto", is_flag = True, help = "Configure node settings manually.")
 @onBeforeCommandExecute(node_module.initializeNodeConfiguration)
-def update() -> None:
+def update(auto: bool) -> None:
```
Comment (dule1322) [inline coretex/cli/commands/node.py:78]: Add "-y" and "-n" flags which define if a prompt will be automatically responded to with a yes or with a no

File: coretex/cli/modules/utils.py [hunk 16]
```diff
@@ -15,22 +15,53 @@
...
+    if DEFAULT_VENV_PATH.exists():
+        return
+
+    venv.create(DEFAULT_VENV_PATH, with_pip = True)
+
+    if platform.system() == "Windows":
+        venvPython = DEFAULT_VENV_PATH / "Scripts" / "python.exe"
+
+    currentCtxVersion = getLibraryVersion("coretex")
+    command([str(venvPython), "-m", "pip", "install", f"coretex=={currentCtxVersion}"], ignoreStdout = True)
```
Comment (dule1322) [inline coretex/cli/modules/utils.py:45]: This will install production CLI from PyPi, what if someone installed it using git url?

Comment (bogdant36) [reply coretex/cli/modules/utils.py:45]: I get the point but should we bother with that now?

File: coretex/cli/modules/utils.py [hunk 17]
```diff
@@ -15,22 +15,64 @@
...
 from py3nvml import py3nvml
 
 import click
+import platform
 
+from ...configuration import DEFAULT_VENV_PATH
 from ...utils.process import command
 
 
+def fetchCoretexVersion() -> Optional[str]:
```
Comment (dule1322) [inline coretex/cli/modules/utils.py:33]: This is not retrieving the version but installation source

File: coretex/cli/modules/utils.py [hunk 18]
```diff
@@ -15,22 +15,64 @@
...
+    _, dockerPath, _ = command(["which", "docker"], ignoreStdout = True, ignoreStderr = True)
+    _, gitPath, _ = command(["which", "git"], ignoreStdout = True, ignoreStderr = True)
+
+    dockerPathParts = dockerPath.strip().split('/')
+    dockerExecPath = '/'.join(dockerPathParts[:-1])
+
+    gitPathParts = gitPath.strip().split('/')
+    gitExecPath = '/'.join(gitPathParts[:-1])
+
+    return dockerExecPath, gitExecPath
```
Comment (dule1322) [inline coretex/cli/modules/utils.py:73]: This should be a function which gets the path to a binary of a single executable. ex. `getExecPath("docker")`, `getExecPath("git")`, but not for both at the same time

**Review by dule1322 — APPROVED** [review]
